src/Compiler/Symbols/FSharpDiagnostic.fsi

```diff
@@ -7,9 +7,99 @@
 namespace FSharp.Compiler.Diagnostics
 
 open System
+open FSharp.Compiler.Symbols
 open FSharp.Compiler.Text
 open FSharp.Compiler.DiagnosticsLogger
 
+/// Information about the context of a type equation in type-mismatch-like diagnostic
+[<RequireQualifiedAccess; Experimental("This FCS API is experimental and subject to change.")>]
+type public DiagnosticContextInfo =
+    /// No context was given
+    | NoContext
+    /// The type equation comes from an IF expression
+    | IfExpression
+    /// The type equation comes from an omitted else branch
+    | OmittedElseBranch
+    /// The type equation comes from a type check of the result of an else branch
+    | ElseBranchResult
+    /// The type equation comes from the verification of record fields
+    | RecordFields
+    /// The type equation comes from the verification of a tuple in record fields
+    | TupleInRecordFields
+    /// The type equation comes from a list or array constructor
+    | CollectionElement
+    /// The type equation comes from a return in a computation expression
+    | ReturnInComputationExpression
+    /// The type equation comes from a yield in a computation expression
+    | YieldInComputationExpression
+    /// The type equation comes from a runtime type test
+    | RuntimeTypeTest
+    /// The type equation comes from an downcast where a upcast could be used
+    | DowncastUsedInsteadOfUpcast
+    /// The type equation comes from a return type of a pattern match clause (not the first clause)
+    | FollowingPatternMatchClause
+    /// The type equation comes from a pattern match guard
+    | PatternMatchGuard
+    /// The type equation comes from a sequence expression
+    | SequenceExpression
+
+/// Contextually-relevant data to each particular diagnostic
+[<Interface; Experimental("This FCS API is experimental and subject to change.")>]
```

Comment:
Not strongly against but not a fan of marker interfaces either. What could be alternative designs here?